### PR TITLE
ResourceServer: Add resource server protobuf and wrapper

### PR DESCRIPTION
**Pull request by ryantxu**
This PR introduces the base package for a "ResourceServer" that extracts the kubernetes wrapper logic from the raw storage logic.  This lets us better iterate the storage logic and can depend on common wrapper support.

When this is solid/stable, we can then remove:
* https://github.com/grafana/grafana/tree/v11.1.0/pkg/apiserver/storage/file
* https://github.com/grafana/grafana/tree/v11.1.0/pkg/services/apiserver/storage/entity

In https://github.com/grafana/grafana/pull/89071 we build a POC using this structure to (almost) support dashboards with arbitrarily large bodies.

In following PRs, we will
- Incorporate the same tests from [storage/file/watcher_test.go
](https://github.com/grafana/grafana/blob/v11.1.0/pkg/apiserver/storage/file/watcher_test.go) https://github.com/grafana/grafana/pull/90161
- Support TTL... just like file store

### CODE REVIEW

File: pkg/apimachinery/utils/meta.go [hunk 1]
```diff
@@ -20,6 +23,8 @@ const AnnoKeyUpdatedTimestamp = "grafana.app/updatedTimestamp"
 const AnnoKeyUpdatedBy = "grafana.app/updatedBy"
 const AnnoKeyFolder = "grafana.app/folder"
 const AnnoKeySlug = "grafana.app/slug"
+const AnnoKeyBlob = "grafana.app/blob"
```
Comment (ryantxu) [inline pkg/apimachinery/utils/meta.go:26]: blob not required for this PR, but including it does not cause any issues and including it here will simplify merge conflicts for the next PR adding blob store suppport

File: pkg/services/apiserver/options/storage.go [hunk 2]
```diff
@@ -4,20 +4,23 @@ import (
...
-	StorageTypeLegacy      StorageType = "legacy"
-	StorageTypeUnified     StorageType = "unified"
-	StorageTypeUnifiedGrpc StorageType = "unified-grpc"
+	StorageTypeFile            StorageType = "file"
+	StorageTypeEtcd            StorageType = "etcd"
+	StorageTypeLegacy          StorageType = "legacy"
+	StorageTypeUnified         StorageType = "unified"
+	StorageTypeUnifiedGrpc     StorageType = "unified-grpc"
+	StorageTypeUnifiedNext     StorageType = "unified-next"
+	StorageTypeUnifiedNextGrpc StorageType = "unified-next-grpc"
```
Comment (ryantxu) [inline pkg/services/apiserver/options/storage.go:23]: Once we are comfortable with this interface running in cloud, we will remove these two options

File: pkg/storage/unified/entitybridge/entitybridge.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+func (b *entityBridge) PrepareList(ctx context.Context, req *resource.ListRequest) (*resource.ListResponse, error) {
+	key := req.Options.Key
+	query := &entity.EntityListRequest{
+		NextPageToken: req.NextPageToken,
+		Limit:         req.Limit,
+		Key:           []string{toEntityKey(key)},
+		WithBody:      true,
+	}
+
+	// Assumes everything is equals
```
Comment (DanCech) [inline pkg/storage/unified/entitybridge/entitybridge.go:247]: we should just pass stuff that is equals, and handle other predicates in storage

File: pkg/storage/unified/resource/cdk_backend.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+	}
+
+	rsp := &ListResponse{}
+	for _, item := range resources {
+		latest := item.versions[0]
+		raw, err := s.bucket.ReadAll(ctx, latest.key)
+		if err != nil {
+			return nil, err
+		}
+		rsp.Items = append(rsp.Items, &ResourceWrapper{
```
Comment (DanCech) [inline pkg/storage/unified/resource/cdk_backend.go:201]: should we be excluding deleted items here?
```suggestion
		if bytes.Contains(raw, []byte(`"DeletedMarker"`)) {
			tmp := &unstructured.Unstructured{}
			err = tmp.UnmarshalJSON(raw)
			if err == nil && tmp.GetKind() == "DeletedMarker" {
				continue
			}
		}
		rsp.Items = append(rsp.Items, &ResourceWrapper{
```

Comment (ryantxu) [reply pkg/storage/unified/resource/cdk_backend.go:201]: good question -- let me improve some tests and find out!

File: pkg/storage/unified/resource/metrics.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	return StorageServerMetrics
+}
+
+func (s *StorageApiMetrics) Collect(ch chan<- prometheus.Metric) {
+	s.OptimisticLockFailed.Collect(ch)
+}
+
+func (s *StorageApiMetrics) Describe(ch chan<- *prometheus.Desc) {
+	s.OptimisticLockFailed.Describe(ch)
+}
```
Comment (DanCech) [inline pkg/storage/unified/resource/metrics.go:41]: Do these make sense? Sooner or later we're going to add another metric, then we'll either need to make more wrappers or switch callers from `s.Collect(ch)` to `s.OptimisticLockFailed.Collect(ch)`

Comment (ryantxu) [reply pkg/storage/unified/resource/metrics.go:41]: This was copied directly from https://github.com/grafana/grafana/blob/main/pkg/services/store/entity/sqlstash/metrics.go#L18

I don't have any opinion on what the best approach is -- we can remove it or replace with something better

File: pkg/storage/unified/resource/resource.proto [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+// ----------------------------------
+// List Request/Response
+// ----------------------------------
+
+// The label filtering requirements:
+// https://github.com/kubernetes/kubernetes/blob/v1.30.1/staging/src/k8s.io/apimachinery/pkg/labels/selector.go#L141
+message Requirement {
+	string key = 1;
+  string operator = 2; // See https://github.com/kubernetes/kubernetes/blob/v1.30.1/staging/src/k8s.io/apimachinery/pkg/selection/operator.go#L21
+	repeated string values = 3; // typically one value, but depends on the operator
```
Comment (DanCech) [inline pkg/storage/unified/resource/resource.proto:159]: something funky here with indentation

File: pkg/storage/unified/resource/resource.proto [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+}
+
+message Sort {
+  enum Order {
+    ASC = 0;
+    DESC = 1;
+  }
+	string field = 1;
+  Order order = 2;
```
Comment (DanCech) [inline pkg/storage/unified/resource/resource.proto:168]: here too

File: pkg/storage/unified/resource/resource.proto (hunk 6 above)
Comment (ryantxu) [reply pkg/storage/unified/resource/resource.proto]: tabs 🙃 spaces

File: pkg/storage/unified/resource/cdk_backend.go (hunk 4 above)
Comment (ryantxu) [reply pkg/storage/unified/resource/cdk_backend.go]: fixed -- thanks.  You are very good at this 😜 !

File: pkg/storage/unified/resource/resource.proto [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+message DeleteResponse {
+  // Status code
+  StatusResult status = 1;
+
+  // The new resource version
+  int64 resource_version = 2;
+
+  // The deleted payload
```
Comment (chaudyg) [inline pkg/storage/unified/resource/resource.proto:128]: I am not sure we need this. Do we ?

File: pkg/storage/unified/resource/resource.proto [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+}
+
+message Sort {
+  enum Order {
+    ASC = 0;
+    DESC = 1;
+  }
+  string field = 1;
+  Order order = 2;
+}
```
Comment (chaudyg) [inline pkg/storage/unified/resource/resource.proto:169]: Not used
```suggestion
```

File: pkg/storage/unified/resource/server.go [hunk 10]
```diff
@@ -0,0 +1,562 @@
...
+	})
+	return err
+}
+
+func (s *server) Watch(req *WatchRequest, srv ResourceStore_WatchServer) error {
+	if err := s.Init(); err != nil {
+		return err
+	}
+
+	fmt.Printf("WATCH %v\n", req.Options.Key)
```
Comment (chaudyg) [inline pkg/storage/unified/resource/server.go:503]:
```suggestion
```

File: pkg/services/apiserver/service.go [hunk 11]
```diff
@@ -258,6 +262,36 @@ func (s *service) start(ctx context.Context) error {
...
+			return fmt.Errorf("unified storage requires the unifiedStorage feature flag")
+		}
+		// Create a connection to the gRPC server
+		conn, err := grpc.NewClient(o.StorageOptions.Address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			return err
+		}
+
+		// TODO: determine when to close the connection, we cannot defer it here
+		// defer conn.Close()
```
Comment (JohnnyQQQQ) [inline pkg/services/apiserver/service.go:288]: I'm not 100% sure about the lifetime of this, but I guess we could use `stoppedCh` in a go routine to close it gracefully close when we exit.

Comment (chaudyg) [reply pkg/services/apiserver/service.go:288]: I am thinking the same. This is copy paste from the previous entity-server code .

File: pkg/storage/unified/resource/resource.proto (hunk 8 above)
Comment (ryantxu) [reply pkg/storage/unified/resource/resource.proto]: I removed it -- since it is an optional return in k8s.  And with protobuf we can always add later :)

**Review by chaudyg — APPROVED** [review]

**Pull request merged**